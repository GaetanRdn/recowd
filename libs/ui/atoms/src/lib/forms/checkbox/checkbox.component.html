<input
  type="checkbox"
  [value]="value"
  [checked]="checked"
  [disabled]="disabled"
/>

<rc-material-icon *ngIf="!checked">check_box_outline_blank</rc-material-icon>
<rc-material-icon *ngIf="checked">check_box</rc-material-icon>

<label>
  <ng-content></ng-content>
</label>
