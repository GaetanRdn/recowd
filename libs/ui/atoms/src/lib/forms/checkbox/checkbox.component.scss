@use '../../assets/utils';

@mixin rc-checkbox($config) {
  $typography: utils.map-get-deep($config, typography, input);
  $colors: map-get($config, colors);
  $primary: utils.map-get-deep($config, colors, primary);

  rc-checkbox {
    @include utils.use-typography($typography);
    position: relative;
    display: inline-flex;
    flex-flow: row nowrap;
    gap: 1.6rem;
    align-items: center;
    cursor: pointer;

    &:is(.rc-disabled) {
      cursor: default;

      > * {
        color: map-get(map-get($colors, disabled), color);
      }
    }

    &.rc-label-before {
      flex-direction: row-reverse;
    }

    label {
      color: map-get($primary, color);
      cursor: inherit;
    }

    input {
      position: absolute;
      left: -9999px;
    }
  }
}
