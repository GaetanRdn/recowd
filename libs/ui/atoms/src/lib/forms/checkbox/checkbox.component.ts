import {
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  EventEmitter,
  forwardRef,
  HostListener,
  Input,
  NgModule,
  Output,
} from '@angular/core';
import { CommonModule } from '@angular/common';
import {
  BooleanInput,
  CoerceBoolean,
  Nullable,
  OnChangeFn,
  OnTouchedFn,
  TypedControlValueAccessor,
} from '@recowd/utility-types';
import { MaterialIconModule } from '../../material-icon/material-icon.component';
import { NG_VALUE_ACCESSOR } from '@angular/forms';

@Component({
  selector: 'rc-checkbox',
  host: {
    '[class.rc-label-before]': 'labelBefore',
    '[class.rc-disabled]': 'disabled',
  },
  templateUrl: './checkbox.component.html',
  providers: [
    {
      provide: NG_VALUE_ACCESSOR,
      useExisting: forwardRef(() => CheckboxComponent),
      multi: true,
    },
  ],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class CheckboxComponent<ValueType>
  implements TypedControlValueAccessor<Nullable<ValueType>>
{
  static ngAcceptInputType_labelBefore: BooleanInput;
  static ngAcceptInputType_checked: BooleanInput;
  static ngAcceptInputType_disabled: BooleanInput;

  @Input() @CoerceBoolean() public labelBefore = false;

  @Input() @CoerceBoolean() public checked = false;
  @Input() public value!: ValueType;
  @Output() public readonly checkedChange: EventEmitter<
    CheckboxComponent<ValueType>
  > = new EventEmitter<CheckboxComponent<ValueType>>();

  constructor(private readonly _cdr: ChangeDetectorRef) {}

  private _disabled = false;

  get disabled(): boolean {
    return this._disabled;
  }

  @Input() @CoerceBoolean() set disabled(disabled: boolean) {
    this._disabled = disabled;
  }

  @HostListener('click')
  public toggle(): void {
    if (!this.disabled) {
      this.checked = !this.checked;
      this.onChange(this.checked ? this.value : null);
      this.onTouched();
      this.checkedChange.emit(this);
    }
  }

  public registerOnChange(fn: OnChangeFn<Nullable<ValueType>>): void {
    this.onChange = fn;
  }

  public registerOnTouched(fn: OnTouchedFn): void {
    this.onTouched = fn;
  }

  public writeValue(value: Nullable<ValueType>): void {
    this.checked = !!value;
  }

  public setDisabledState(isDisabled: boolean): void {
    this._disabled = isDisabled;
    this._cdr.markForCheck();
  }

  private onChange: OnChangeFn<Nullable<ValueType>> = () => {
    // default method
  };

  private onTouched: OnTouchedFn = () => {
    // default method
  };
}

@NgModule({
  declarations: [CheckboxComponent],
  exports: [CheckboxComponent],
  imports: [CommonModule, MaterialIconModule],
})
export class CheckboxModule {}
