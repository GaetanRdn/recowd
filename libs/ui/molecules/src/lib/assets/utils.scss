@function map-get-deep($map, $level, $subLevel) {
  @return map-get($map, $level, $subLevel);
}

@mixin use-typography($typography) {
  @each $key in font-size, font-family, font-weight, line-height {
    #{$key}: map-get($typography, $key);
  }
}
