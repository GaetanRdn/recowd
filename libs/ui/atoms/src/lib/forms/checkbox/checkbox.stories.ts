import { BADGE } from '@geometricpanda/storybook-addon-badges';
import { Meta, moduleMetadata, Story } from '@storybook/angular';
import { CheckboxComponent } from './checkbox.component';
import { action } from '@storybook/addon-actions';
import { ReactiveFormsModule, UntypedFormControl } from '@angular/forms';

export default {
  title: 'atoms/forms/checkbox',
  parameters: {
    badges: [BADGE.STABLE],
    jest: 'checkbox.component',
  },
  decorators: [
    moduleMetadata({
      imports: [CheckboxComponent, ReactiveFormsModule],
    }),
  ],
} as Meta<CheckboxComponent<number>>;

const template: Story<CheckboxComponent<number>> = (
  args: CheckboxComponent<number>
) => ({
  props: {
    labelBefore: args.labelBefore,
    checked: args.checked,
    disabled: args.disabled,
    log: action('checkedChange'),
  },
  template: `<rc-checkbox [labelBefore]="labelBefore" [checked]="checked" [value]="1" [disabled]="disabled" (checkedChange)="log($event)">Used</rc-checkbox>`,
});

export const Default = template.bind({});
Default.args = {
  labelBefore: false,
  checked: false,
  disabled: false,
};

const reactiveTemplate: Story<CheckboxComponent<number>> = (
  args: CheckboxComponent<number>
) => ({
  props: {
    labelBefore: args.labelBefore,
    control: new UntypedFormControl({
      value: args.checked ? 1 : null,
      disabled: args.disabled,
    }),
    log: action('checkedChange'),
  },
  template: `<rc-checkbox [labelBefore]="labelBefore" [formControl]="control" [value]="1" (checkedChange)="log($event)">Reactively Used</rc-checkbox>`,
});

export const Reactive = reactiveTemplate.bind({});
Reactive.args = {
  ...Default.args,
};
