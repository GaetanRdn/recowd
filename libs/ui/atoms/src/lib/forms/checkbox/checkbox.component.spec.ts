import { TemplateLookup } from '@recowd/test/utils';
import { Component, DebugElement } from '@angular/core';
import { TestBed, waitForAsync } from '@angular/core/testing';
import { CheckboxComponent } from './checkbox.component';
import { FormControl, ReactiveFormsModule } from '@angular/forms';

describe('CheckboxComponent', () => {
  beforeEach(waitForAsync(() =>
    TestBed.configureTestingModule({
      declarations: [HostComponent, ReactiveHostComponent],
      imports: [CheckboxComponent, ReactiveFormsModule],
    }).compileComponents()));

  describe('Basic', () => {
    let templateLookup: TemplateLookup<HostComponent>;

    beforeEach(() => (templateLookup = new TemplateLookup(HostComponent)));

    test('should create', () => {
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('should create with label before', () => {
      templateLookup.hostComponent.labelBefore = true;
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('should create checked', () => {
      templateLookup.hostComponent.checked = true;
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('should create disabled', () => {
      templateLookup.hostComponent.disabled = true;
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('should not change when disabled', () => {
      // GIVEN
      jest.spyOn(templateLookup.hostComponent, 'checkedChange');
      templateLookup.hostComponent.disabled = true;
      templateLookup.detectChanges();

      // WHEN
      templateLookup.query('rc-checkbox').click();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(templateLookup.hostComponent.checkedChange).not.toBeCalled();
    });

    test.each([false, true])(
      'should toggle checked on click (was checked %s)',
      (wasChecked: boolean) => {
        // GIVEN
        jest.spyOn(templateLookup.hostComponent, 'checkedChange');
        templateLookup.hostComponent.checked = wasChecked;
        templateLookup.detectChanges();

        // WHEN
        const checkbox: DebugElement = templateLookup.get(CheckboxComponent);
        checkbox.nativeElement.click();
        templateLookup.detectChanges();

        // THEN
        expect(templateLookup.firstChildElement).toMatchSnapshot();
        expect(templateLookup.hostComponent.checkedChange).toBeCalledWith(
          checkbox.componentInstance
        );
      }
    );
  });

  describe('Reactive form', () => {
    let templateLookup: TemplateLookup<ReactiveHostComponent>;

    beforeEach(
      () => (templateLookup = new TemplateLookup(ReactiveHostComponent))
    );

    test('should create', () => {
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('should create checked', () => {
      // GIVEN
      templateLookup.hostComponent.control.setValue('toto');

      // WHEN
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('should create disabled', () => {
      // GIVEN
      templateLookup.hostComponent.control.disable();

      // WHEN
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('should update control on click', () => {
      // GIVEN
      templateLookup.detectChanges();

      // WHEN
      templateLookup.query('rc-checkbox').click();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(templateLookup.hostComponent.control.value).toEqual('toto');
    });

    test('should set control value to null on click', () => {
      // GIVEN
      templateLookup.hostComponent.control.setValue('toto');
      templateLookup.detectChanges();

      // WHEN
      templateLookup.query('rc-checkbox').click();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(templateLookup.hostComponent.control.value).toBeNull();
    });
  });
});

@Component({
  template: ` <rc-checkbox
    [labelBefore]="labelBefore"
    [checked]="checked"
    [value]="value"
    [disabled]="disabled"
    (checkedChange)="checkedChange($event)"
    >Label
  </rc-checkbox>`,
})
class HostComponent {
  public labelBefore = false;
  public checked = false;
  public disabled = false;
  public value = 'val';

  // eslint-disable-next-line @typescript-eslint/no-unused-vars,@typescript-eslint/no-empty-function
  public checkedChange(_: unknown): void {}
}

@Component({
  template: ` <rc-checkbox [formControl]="control" [value]="'toto'"
    >Label
  </rc-checkbox>`,
})
class ReactiveHostComponent {
  public readonly control: FormControl = new FormControl();
}
