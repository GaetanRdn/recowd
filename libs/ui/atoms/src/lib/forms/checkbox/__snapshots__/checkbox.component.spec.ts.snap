// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxComponent Basic should create 1`] = `
<rc-checkbox>
  <input
    type="checkbox"
    value="val"
  />
  <rc-material-icon
    class="material-icons"
  >
    check_box_outline_blank
  </rc-material-icon>
  
  
  <label>
    Label 
  </label>
</rc-checkbox>
`;

exports[`CheckboxComponent Basic should create checked 1`] = `
<rc-checkbox>
  <input
    type="checkbox"
    value="val"
  />
  
  <rc-material-icon
    class="material-icons"
  >
    check_box
  </rc-material-icon>
  
  <label>
    Label 
  </label>
</rc-checkbox>
`;

exports[`CheckboxComponent Basic should create disabled 1`] = `
<rc-checkbox
  class="rc-disabled"
>
  <input
    disabled=""
    type="checkbox"
    value="val"
  />
  <rc-material-icon
    class="material-icons"
  >
    check_box_outline_blank
  </rc-material-icon>
  
  
  <label>
    Label 
  </label>
</rc-checkbox>
`;

exports[`CheckboxComponent Basic should create with label before 1`] = `
<rc-checkbox
  class="rc-label-before"
>
  <input
    type="checkbox"
    value="val"
  />
  <rc-material-icon
    class="material-icons"
  >
    check_box_outline_blank
  </rc-material-icon>
  
  
  <label>
    Label 
  </label>
</rc-checkbox>
`;

exports[`CheckboxComponent Basic should not change when disabled 1`] = `
<rc-checkbox
  class="rc-disabled"
>
  <input
    disabled=""
    type="checkbox"
    value="val"
  />
  <rc-material-icon
    class="material-icons"
  >
    check_box_outline_blank
  </rc-material-icon>
  
  
  <label>
    Label 
  </label>
</rc-checkbox>
`;

exports[`CheckboxComponent Basic should toggle checked on click (was checked false) 1`] = `
<rc-checkbox>
  <input
    type="checkbox"
    value="val"
  />
  
  <rc-material-icon
    class="material-icons"
  >
    check_box
  </rc-material-icon>
  
  <label>
    Label 
  </label>
</rc-checkbox>
`;

exports[`CheckboxComponent Basic should toggle checked on click (was checked true) 1`] = `
<rc-checkbox>
  <input
    type="checkbox"
    value="val"
  />
  <rc-material-icon
    class="material-icons"
  >
    check_box_outline_blank
  </rc-material-icon>
  
  
  <label>
    Label 
  </label>
</rc-checkbox>
`;

exports[`CheckboxComponent Reactive form should create 1`] = `
<rc-checkbox
  class="ng-untouched ng-pristine ng-valid"
>
  <input
    type="checkbox"
    value="toto"
  />
  <rc-material-icon
    class="material-icons"
  >
    check_box_outline_blank
  </rc-material-icon>
  
  
  <label>
    Label 
  </label>
</rc-checkbox>
`;

exports[`CheckboxComponent Reactive form should create checked 1`] = `
<rc-checkbox
  class="ng-untouched ng-pristine ng-valid"
>
  <input
    type="checkbox"
    value="toto"
  />
  
  <rc-material-icon
    class="material-icons"
  >
    check_box
  </rc-material-icon>
  
  <label>
    Label 
  </label>
</rc-checkbox>
`;

exports[`CheckboxComponent Reactive form should create disabled 1`] = `
<rc-checkbox
  class="rc-disabled ng-untouched ng-pristine"
>
  <input
    disabled=""
    type="checkbox"
    value="toto"
  />
  <rc-material-icon
    class="material-icons"
  >
    check_box_outline_blank
  </rc-material-icon>
  
  
  <label>
    Label 
  </label>
</rc-checkbox>
`;

exports[`CheckboxComponent Reactive form should set control value to null on click 1`] = `
<rc-checkbox
  class="ng-valid ng-touched ng-dirty"
>
  <input
    type="checkbox"
    value="toto"
  />
  <rc-material-icon
    class="material-icons"
  >
    check_box_outline_blank
  </rc-material-icon>
  
  
  <label>
    Label 
  </label>
</rc-checkbox>
`;

exports[`CheckboxComponent Reactive form should update control on click 1`] = `
<rc-checkbox
  class="ng-valid ng-touched ng-dirty"
>
  <input
    type="checkbox"
    value="toto"
  />
  
  <rc-material-icon
    class="material-icons"
  >
    check_box
  </rc-material-icon>
  
  <label>
    Label 
  </label>
</rc-checkbox>
`;
